p2p/src/types.rs

```diff
@@ -0,0 +1,473 @@
+    ///
+    /// Note: For `Dns` addresses, private IP checks are performed after resolution in
+    /// [`resolve_filtered`](Self::resolve_filtered).
+    pub fn is_valid(&self, allow_private_ips: bool, allow_dns: bool) -> bool {
+        match self {
+            Self::Socket(addr) => allow_private_ips || IpAddrExt::is_global(&addr.ip()),
+            Self::Dns { .. } => allow_dns,
+        }
+    }
+
+    /// Resolve this ingress address to a socket address.
+    ///
+    /// For `Socket` variants, returns the address directly.
+    /// For `Dns` variants, performs DNS resolution and returns the first resolved IP.
+    pub async fn resolve(&self, resolver: &impl Resolver) -> Result<SocketAddr, RuntimeError> {
+        match self {
+            Self::Socket(addr) => Ok(*addr),
+            Self::Dns { host, port } => {
+                let ips = resolver.resolve(host.as_str()).await?;
+                let ip = ips
+                    .first()
+                    .ok_or_else(|| RuntimeError::ResolveFailed(host.to_string()))?;
+                Ok(SocketAddr::new(*ip, *port))
+            }
+        }
+    }
+
+    /// Resolve and filter by private IP policy.
+    ///
+    /// Returns `None` if:
+    /// - DNS resolution fails
+    /// - The resolved IP is private and `allow_private_ips` is false
+    pub async fn resolve_filtered(
+        &self,
+        resolver: &impl Resolver,
+        allow_private_ips: bool,
+    ) -> Option<SocketAddr> {
+        let addr = self.resolve(resolver).await.ok()?;
+        if !allow_private_ips && !IpAddrExt::is_global(&addr.ip()) {
+            return None;
+        }
+        Some(addr)
+    }
+}
+
+impl Write for Ingress {
+    fn write(&self, buf: &mut impl BufMut) {
+        match self {
+            Self::Socket(addr) => {
+                INGRESS_SOCKET_PREFIX.write(buf);
+                addr.write(buf);
+            }
+            Self::Dns { host, port } => {
+                INGRESS_DNS_PREFIX.write(buf);
+                host.write(buf);
+                port.write(buf);
+            }
+        }
+    }
+}
+
+impl EncodeSize for Ingress {
+    fn encode_size(&self) -> usize {
+        u8::SIZE
+            + match self {
+                Self::Socket(addr) => addr.encode_size(),
+                Self::Dns { host, port } => host.encode_size() + port.encode_size(),
+            }
+    }
+}
+
+impl Read for Ingress {
+    type Cfg = ();
+
+    fn read_cfg(buf: &mut impl Buf, _cfg: &Self::Cfg) -> Result<Self, CodecError> {
+        let prefix = u8::read(buf)?;
+        match prefix {
+            INGRESS_SOCKET_PREFIX => {
+                let addr = SocketAddr::read(buf)?;
+                Ok(Self::Socket(addr))
+            }
+            INGRESS_DNS_PREFIX => {
+                let host = Hostname::read(buf)?;
+                let port = u16::read(buf)?;
+                Ok(Self::Dns { host, port })
+            }
+            _ => Err(CodecError::Invalid("Ingress", "Invalid prefix")),
```

Comment:
I think we typically use `CodecError::InvalidEnum`

...arguably it should come with a `&'static str` to give more context but it only takes in the u8 value for now